src/crimson/osd/exceptions.h

```diff
@@ -60,3 +60,525 @@ struct input_output_error : public error {
+
+  public:
+    using errorator_type = ceph::errorator<WrappedAllowedErrorsT...>;
+
+    future(base_t&& base)
+      : base_t(std::move(base)) {
+    }
+
+    template <template <class...> class ErroratedFuture,
+              class = std::void_t<
+                typename ErroratedFuture<ValuesT...>::errorator_type>>
+    operator ErroratedFuture<ValuesT...> () && {
+      using DestErrorator = \
+        typename ErroratedFuture<ValuesT...>::errorator_type;
+      using ThisErrorator = errorator<WrappedAllowedErrorsT...>;
+
+      static_assert(!DestErrorator::template is_less_errorated_v<ThisErrorator>,
+                    "conversion is possible to more-or-eq errorated future!");
+      return plainify(ErroratedFuture<ValuesT...>(std::move(*this)));
+    }
+
+    // initialize future as failed without throwing. `make_exception_future()`
+    // internally uses `std::make_exception_ptr()`. cppreference.com shouldn't
+    // be misinterpreted when it says:
+    //
+    //   "This is done as if executing the following code:
+    //     try {
+    //         throw e;
+    //     } catch(...) {
+    //         return std::current_exception();
+    //     }",
+    //
+    // the "as if" is absolutely crucial because modern GCCs employ optimized
+    // path for it. See:
+    //   * https://gcc.gnu.org/git/?p=gcc.git;a=commit;h=cce8e59224e18858749a2324bce583bcfd160d6c,
+    //   * https://gcc.gnu.org/ml/gcc-patches/2016-08/msg00373.html.
+    //
+    // This behavior, combined with `__cxa_exception_type()` for inspecting
+    // exception's type, allows for throw/catch-free handling of stateless
+    // exceptions (which is fine for error codes). Stateful jumbos would be
+    // actually a bit harder as `_M_get()` is private, and thus rethrowing is
+    // necessary to get to the state inside. However, it's not unthinkable to
+    // see another extension bringing operator*() to the exception pointer...
+    //
+    // TODO: we don't really need to `make_exception_ptr` each time. It still
+    // allocates memory underneath while can be replaced with single instance
+    // per type created on start-up.
+    template <class ErrorT,
+              class DecayedT = std::decay_t<ErrorT>,
+              bool IsAllowed = (std::is_same_v<WrappedAllowedErrorsT,
+                                               DecayedT> || ...),
+              class = std::enable_if_t<IsAllowed>>
+    future(ErrorT&& e)
+      : base_t(
+          seastar::make_exception_future<ValuesT...>(
+            errorator<WrappedAllowedErrorsT...>::make_exception_ptr(e))) {
+      // this is `fold expression` of C++17. Because of the SFINAE above
+      // the assertion cannot fail. Keeping it just in the case.
+      static_assert((... || (std::is_same_v<DecayedT, WrappedAllowedErrorsT>)),
+                    "ErrorT is not enlisted in errorator");
+    }
+
+    template <class ValueFuncT, class ErrorVisitorT>
+    auto safe_then(ValueFuncT&& valfunc, ErrorVisitorT&& errfunc) {
```

Comment:
If I've got this right, errfunc must be exaustive.  In many cases, it would be nice to be able to handle some, but not all errors and propogate a subset (or w/e) errorator out of a function.  Perhaps we ought to call this one safe_then_exhaustive and later introduce a safe_then which will propogate an errorated future with error set comprised of the union of the unhandled errors from the original errorator with the errors from the safe_then result branches?

Reply:
No, no. It should be possible to use `pass_further<>` for a subset. In my working branch I'm doing something similar but with `discard_all`:

```cpp
  case CEPH_OSD_OP_CALL:
    return this->do_op_call(osd_op).safe_then(
      [] {
        return seastar::now();
      }, ceph::stateful_errint::handle([] (int err) {
        // TODO: implement the handler. NOP for now.
      }), ceph::ct_error::input_output_error::handle([] {
        // TODO: implement the handler. NOP for now.
      }),
      ceph::errorator<ceph::ct_error::enoent,
                      ceph::ct_error::operation_not_supported,
                      ceph::ct_error::permission_denied,
                      ceph::ct_error::invarg>::discard_all{}
      );
```

Reply:
I suppose that would work too.  With a safe_then chain, though, you'd have to repeatedly specify which ones you're leaving to be handled later.  I'd rather only have to deal with that at function type boundaries.

Reply:
> you'd have to repeatedly specify which ones you're leaving to be handled later

Yes, that's correct. Personally I don't know whether this is an advantage or disadvantage. Looks like a trade-off between easiness-of-use and error-proneness.

The `ceph::pass_further<error1, error2>{}` mentioned earlier is another face of the same dilemma, I believe.

Reply:
I don't think there's much of an error-proneness problem since the function signature boundaries will still need to use a type which correctly enumerates the allowed errors.  I think I'd prefer the safe_then with implicit pass_further for ease of chaining.